app/src/main/java/cloudcity/LoggingServiceExtensions.java

```diff
@@ -0,0 +1,214 @@
+
+import androidx.annotation.NonNull;
+
+import java.util.List;
+import java.util.Objects;
+
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.DataProvider.CellInformations.CellInformation;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.DataProvider.CellInformations.GSMInformation;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.DataProvider.CellInformations.LTEInformation;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.DataProvider.CellInformations.NRInformation;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.DataProvider.DataProvider;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.DataProvider.LocationInformation;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.GlobalVars;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Preferences.SPType;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.Preferences.SharedPreferencesGrouper;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.cloudCity.CloudCityHelpers;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.cloudCity.models.CellInfoModel;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.cloudCity.models.MeasurementsModel;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.cloudCity.models.NetworkDataModel;
+import de.fraunhofer.fokus.OpenMobileNetworkToolkit.cloudCity.models.NetworkDataModelRequest;
+
+public class LoggingServiceExtensions {
+    private static final String TAG = "LoggingServiceExtensions";
+
+    private static Handler CloudCityHandler;
+    private static SharedPreferences sp;
+
+    private static GlobalVars gv;
+
+    private static int interval;
+
+    private static DataProvider dp;
+
+    // Handle remote Cloud City update
+    private final static Runnable CloudCityUpdate = new Runnable() {
+        @Override
+        public void run() {
+            try {
+                Log.d(TAG, "run: CC Update");
+
+                String address = sp.getString("cloud_city_url", "");
+                String token = sp.getString("cloud_city_token", "");
+
+                NetworkDataModel data = getCloudCityData();
+                if (data == null) {
+                    Log.e(TAG, "run: Error in getting data from Cloud city, skipping sending");
+                    return;
+                }
+                NetworkDataModelRequest requestData = new NetworkDataModelRequest();
+                requestData.add(data);
+
+                boolean status = CloudCityHelpers.sendData(address, token, requestData);
+
+                if (status) {
+                    /* Data sent successfully indicate in status icon. */
+                    gv.getLog_status().setColorFilter(Color.argb(255, 0, 255, 0));
+                } else  {
+                    gv.getLog_status().setColorFilter(Color.argb(255, 255, 0, 0));
+                }
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+
+            CloudCityHandler.postDelayed(this, interval);
+        }
+    };
```

Comment:
this is bad, Rather than using their SharedPreferenceGrouper, why don't we just make a CloudCityParamRepository to hold the cloud city server URL and token, and save it directly to SharedPref using a "normal" plain shared pref instead of going through their grouper and saving in default one.

And when we initialize the Repo in the Application::onCreate, we'll ensure that it's prefilled with proper data before it gets to MainActivity and all subsequent screens.